### PR TITLE
chore(buildDockerAndPublishImage) cleanup code and tests

### PR DESCRIPTION
**Pull request by dduportal**
This PR aims at cleaning up the code used for the Docker(s) shared libraries.

It introduces the following changes:

- Deprecate unused arguments for `buildDockerAndPublishImage()` (`mainBranch`, `registry`, `metadataFromSh`)
- Use the environment variable `env.BRANCH_IS_PRIMARY` (instead of specifying a principal branch that adds unused code)
- Cleanup test code of `buildDockerAndPublishImage`  based on recent @lemeurherve feedbacks about repetition (there still have some to do but it's a first step). Includes deleting old tests or tests that made no sense.
- Updating some `*.txt` documentation (as HTML is not parsed anymore)
- Removing the `src/*/Docker` class and switch to simple map merging for default attributes
- Set the expected default value of `credentials` (used for Docker engine/`img`) in `buildDockerAndPublishImage()` to our default value

### CODE REVIEW

File: vars/buildDockerAndPublishImage.txt [hunk 1]
```diff
@@ -1,87 +1,63 @@
-<p>
-    Lints, Builds, Tests and eventually Publishes a Docker image.
+Lints, Builds, Tests and eventually Deploys a Docker image.
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:1]:
```suggestion
Lints, builds, tests and eventually deploys a Docker image.
```

File: vars/buildDockerAndPublishImage.txt [hunk 2]
```diff
@@ -1,87 +1,63 @@
...
-        </ul>
-      </li>
-    </ul>
-    </p>
+The following arguments are available for this function:
 
-    <p>
-    The Lint phase generates a report when it fails, recorded by the hadolint tool in your Jenkins instance.
-    </p>
+- String **imageName**: (Mandatory) name of the image to built, usually referenced as the "Repository" part of a Docker image's name without any tag (Example: "builder" or "terraform").
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:7]:
```suggestion
- String **imageName**: (Mandatory) name of the image to build, usually referenced as the "Repository" part of a Docker image's name without any tag (Example: "builder" or "terraform").
```

File: vars/buildDockerAndPublishImage.txt [hunk 3]
```diff
@@ -1,87 +1,63 @@
...
+- Map **config**: (Optional) custom configuration for the image build an deploy process. The maps can holds any of the following keys:
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:8]:
```suggestion
- Map **config**: (Optional) custom configuration for the image build and deploy process. The maps can holds any of the following keys:
```
(?)

File: vars/buildDockerAndPublishImage.txt [hunk 4]
```diff
@@ -1,87 +1,63 @@
...
+  * String **builderImage**: (Optional, defaults to "jenkinsciinfra/builder:2.0.2") Docker Image to be used for the container "builder" of the CI pod template.
+  * String **platform**: (Optional, defaults to "linux/amd64") Name of the docker platform to use when building this image, for multiple use a comma separated list (Example: "linux/amd64,linux/arm64").
+  * Boolean **automaticSemanticVersioning**: (Optional, defaults to "false") Should we create a release for every merge to the mainBranch. This uses "jx-release-version" to determine the version number based on the commit history.
+  * String **gitCredentials**: (Optional, defaults to "") If "automaticSemanticVersioning" is set, name of the credential to use when tagging the git repository. Support User/password or GitHub App credential types
+  * String **nextVersionCommand** (Optional, defaults to "jx-release-version" ) If "automaticSemanticVersioning" is set, this is the command to retrieve the next version (to use for tagging and releasing)
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:16]:
```suggestion
  * String **nextVersionCommand** (Optional, defaults to "jx-release-version") If "automaticSemanticVersioning" is set, this is the command to retrieve the next version (to use for tagging and releasing)
```

File: vars/buildDockerAndPublishImage.txt [hunk 5]
```diff
@@ -1,87 +1,63 @@
...
 
-    <p>
-    The build phase adds a bunch of build args that you can use in your Dockerfile:
-    <ul>
-      <li>GIT_COMMIT_REV - The commit that triggered this build</li>
-      <li>GIT_SCM_URL - Url to repo </li>
-      <li>BUILD_DATE - Date that the image was built (now) </li>
-    </ul>
-    </p>
+The Lint phase generates a report when it fails, recorded by the hadolint tool in your Jenkins instance.
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:18]:
```suggestion
The lint phase generates a report when it fails, recorded by the hadolint tool in your Jenkins instance.
```

File: vars/buildDockerAndPublishImage.txt [hunk 6]
```diff
@@ -1,87 +1,63 @@
...
 
-    <p>
-    The test phase is only executed if a file `cst.yml` is found at the root of the repository with the "TAR" driver for security concerns.
-    Please reference to <a href="https://github.com/GoogleContainerTools/container-structure-test">Google's Container Structure Tests documentation</a>.
-    </p>
+The build phase adds a bunch of build args that you can use in your Dockerfile:
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:20]:
```suggestion
The build phase adds build args that you can use in your Dockerfile:
```

File: vars/buildDockerAndPublishImage.txt [hunk 7]
```diff
@@ -1,87 +1,63 @@
...
 
-    <p>
-    Publication of the Docker Image only happens when the build is triggered by:
-    <ul>
-      <li>A push event on the repository's principal branch referenced by "config.mainBranch".</li>
-      <li>A tag event on the repository.</li>
-    </ul>
+- **GIT_COMMIT_REV** - The commit that triggered this build
+- **GIT_SCM_URL** - Url to repo
+- **BUILD_DATE** - Date that the image was built (now)
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:24]:
```suggestion
- **BUILD_DATE** - Date of the image build (now)
```

File: vars/buildDockerAndPublishImage.txt [hunk 8]
```diff
@@ -1,87 +1,63 @@
...
+Please reference to https://github.com/GoogleContainerTools/container-structure-test.
 
-    <pre><code>
-    buildDockerAndPublishImage('terraform', [
-      mainBranch: 'main',
-      dockerfile: 'docker/Dockerfile',
-      registry: 'docker4eval'
-    ])
-    </code></pre>
+Publication of the Docker Image only happens when the build is triggered by:
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:29]:
```suggestion
Publication of the Docker image only happens when the build is triggered by:
```

File: vars/buildDockerAndPublishImage.txt [hunk 9]
```diff
@@ -1,87 +1,63 @@
...
+
+- A push event on the repository's principal branch referenced by "config.mainBranch".
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:31]:
```suggestion
- A push event on the repository's principal branch.
```

**Review by lemeurherve — APPROVED** [review]
Nitpicks only, LGTM :)

File: vars/parallelDockerUpdatecli.txt [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
-- String **mainBranch**: (Optional - Default: "main") name of the main branch of the repository
 - String **rebuildImageOnPeriodicJob**: (Optional - Default: true) specify if the docker image has to be rebuilt and republished when triggered by a periodic job.
 - String **updatecliApplyCronTriggerExpression**: (Optional - Default: "@weekly") Enable periodic updatecli execution by providing a cron-like expression.
 - String **credentialsId**: (Optional - Default: "updatecli-github-token") specify the credentials used by updatecli to interact with github.
 - Map **updatecliConfig** (Optional - Default: empty map) specify custom parameters for the underlying "updatecli()" actions. Please look at the "updatecli()" global library documentation.
+
 [Deprecated]
 - String **containerMemory**: (Optional - Default: "128Mi") specify the amount of memory dedicated to the updatecli container.
+- String **mainBranch**: (Optional - Default: "main") name of the main branch of the repository
+
```
Comment (lemeurherve) [inline vars/parallelDockerUpdatecli.txt:14]:
```suggestion
```

**Review by lemeurherve — DISMISSED** [review]

**Comment by dduportal** [comment]
(putting as draft: test in progress in https://github.com/jenkins-infra/docker-jenkins-weekly/pull/404)

**Comment by dduportal** [comment]
Ready to review!

**Comment by dduportal** [comment]
@lemeurherve your suggestions had been applied (marked as outdated). My last push force might have lost the changes, sorry for that

File: vars/buildDockerAndPublishImage.txt [hunk 11]
```diff
@@ -1,87 +1,63 @@
...
-    <pre><code>
-    buildDockerAndPublishImage('terraform', [
-      mainBranch: 'main',
-      dockerfile: 'docker/Dockerfile',
-      registry: 'docker4eval'
-    ])
-    </code></pre>
+Publication of the Docker image only happens when the build is triggered by:
+
+- A push event on the repository's principal branch.
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:31]:
```suggestion
- A push event on the repository principal branch.
```

File: vars/buildDockerAndPublishImage.txt [hunk 12]
```diff
@@ -1,87 +1,63 @@
...
-        </ul>
-      </li>
-    </ul>
-    </p>
+The following arguments are available for this function:
 
-    <p>
-    The Lint phase generates a report when it fails, recorded by the hadolint tool in your Jenkins instance.
-    </p>
+- String **imageName**: (Mandatory) name of the image to build, usually referenced as the "Repository" part of a Docker image's name without any tag (Example: "builder" or "terraform").
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:7]:
```suggestion
- String **imageName**: (Mandatory) name of the image to build, usually referenced as the "Repository" part of a Docker image name without any tag (Example: "builder" or "terraform").
```

File: vars/buildDockerAndPublishImage.txt [hunk 13]
```diff
@@ -1,87 +1,63 @@
...
+- Map **config**: (Optional) custom configuration for the image build and deploy process. The maps can holds any of the following keys:
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:8]:
```suggestion
- Map **config**: (Optional) custom configuration for the image build and deploy process. This map can holds any of the following keys:
```

File: vars/buildDockerAndPublishImage.txt [hunk 14]
```diff
@@ -1,87 +1,63 @@
...
+  * String **dockerfile**: (Optional, defaults to "Dockerfile") Relative path to the Dockerfile to use within the repository (Example: "build.dockerfile", "docker/Dockerfile").
+  * String **imageDir**: (Optional, defaults to the parent directory of "Dockerfile") Path to a directory to use as build context (Example: "docker/", "python/2.7").
+  * String **credentials**: (Optional, defaults to "jenkins-dockerhub") Name of the Jenkins' credentials used to authenticate on the Docker remote registry during the full build process. (Example: "internal-registry-credential").
+  * String **builderImage**: (Optional, defaults to "jenkinsciinfra/builder:2.0.2") Docker Image to be used for the container "builder" of the CI pod template.
+  * String **platform**: (Optional, defaults to "linux/amd64") Name of the docker platform to use when building this image, for multiple use a comma separated list (Example: "linux/amd64,linux/arm64").
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:13]:
```suggestion
  * String **platform**: (Optional, defaults to "linux/amd64") Name of the docker platform to use when building this image. For multiple platforms use a comma separated list (Example: "linux/amd64,linux/arm64").
```

File: vars/buildDockerAndPublishImage.txt [hunk 15]
```diff
@@ -1,87 +1,63 @@
...
+  * Boolean **automaticSemanticVersioning**: (Optional, defaults to "false") Should we create a release for every merge to the mainBranch. This uses "jx-release-version" to determine the version number based on the commit history.
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:14]:
```suggestion
  * Boolean **automaticSemanticVersioning**: (Optional, defaults to "false") Should a release be created for every merge to the mainBranch. This uses "jx-release-version" to determine the version number based on the commit history.
```

File: vars/buildDockerAndPublishImage.txt [hunk 16]
```diff
@@ -1,87 +1,63 @@
...
+  * String **gitCredentials**: (Optional, defaults to "") If "automaticSemanticVersioning" is set, name of the credential to use when tagging the git repository. Support User/password or GitHub App credential types
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:15]:
```suggestion
  * String **gitCredentials**: (Optional, defaults to "") If "automaticSemanticVersioning" is set, name of the credential to use when tagging the git repository. Support user/password or GitHub App credential types.
```

File: vars/buildDockerAndPublishImage.txt [hunk 17]
```diff
@@ -1,87 +1,63 @@
...
-      <li>BUILD_DATE - Date that the image was built (now) </li>
-    </ul>
-    </p>
+The lint phase generates a report when it fails, recorded by the hadolint tool in your Jenkins instance.
 
-    <p>
-    The test phase is only executed if a file `cst.yml` is found at the root of the repository with the "TAR" driver for security concerns.
-    Please reference to <a href="https://github.com/GoogleContainerTools/container-structure-test">Google's Container Structure Tests documentation</a>.
-    </p>
+The build phase adds build args that you can use in your Dockerfile:
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:20]:
```suggestion
The build phase adds build args you can use in your Dockerfile:
```

File: vars/buildDockerAndPublishImage.txt [hunk 18]
```diff
@@ -1,87 +1,63 @@
...
-    buildDockerAndPublishImage('terraform', [
-      mainBranch: 'main',
-      dockerfile: 'docker/Dockerfile',
-      registry: 'docker4eval'
-    ])
-    </code></pre>
+Publication of the Docker image only happens when the build is triggered by:
+
+- A push event on the repository's principal branch.
+- A tag event on the repository.
```
Comment (lemeurherve) [inline vars/buildDockerAndPublishImage.txt:32]:
```suggestion
- A tag event on the repository
```

**Pull request merged**